Add Seasonal backgrounds

hbnrmx opened the pull request:
Resolves #7055

PRing this as a first draft because I would like some assistance with the remaining issues.

1. stable gives the user the option Seasonal Backgrounds: `Always` / `Sometimes` (default) / `Never`
Not sure if this is worth keeping around in lazers Background source. For now, I only added `Seasonal` (default) 

1. we definitely need some way of caching or prefetching, but I would like some guidance on which level to do that.

1. In `BackgroundScreenDefault` I had to change the initial load to `LoadComponentAsync`
 because of `[LongRunningLoad]` but chaining `LoadComponentAsync`s like that looks very wrong to me.

Commit:
Add BackgroundSource.Seasonal

Commit:
Set BackgroundSource.Seasonal as default setting

Commit:
Add GetSeasonalBackgroundsRequest

Commit:
Add SeasonalBackgroundLoader and SeasonalBackground

Commit:
Load SeasonalBackgroundLoader asynchronously

bdach commented:
No offence, but I think it would have been wiser to ask your questions in an issue comment. Fetching the images every time is pretty shocking.

Here's a proposal of a completely different flow instead:

* On game load, check if there are seasonal backgrounds available
* If there are new backgrounds, fetch them all at once in the background and persist them to storage, then display a notification to the user prompting if they would like to switch
* If the backgrounds went away, clean up the old ones and restore the user's setting to the one before going to seasonal

Could also have seasonal as a separate option, I guess. It can't be as done currently as every switch to seasonal will then require users to revert back manually.

peppy commented:
All at once sounds like not a great idea. Stable currently downloads each time a new one is required asynchronously, and caches after retrieval.

bdach commented:
I guess it's a pretty huge footprint, true. The seasonal background API response should probably persist across the current session, though, to avoid re-querying to check if there is a need to fetch a new one (`SessionStatics`?)

hbnrmx commented:
For reference, the current Halloween event has 15 images, so ~6Mb.
`SessionStatics` only stores boolean configuration values currently, so it may not be built with textures in mind.
That said, I don't know half a thing about caching...

bdach commented:
I was suggesting storing the API response there, not the textures themselves. For textures, the framework `LargeTextureStore` already does some caching, but I believe it won't have much of an effect here as the textures are refcounted - so if all references to a background are lost the texture will get unloaded, which I think will happen on screen transitions rendering the caching pretty pointless in this particular case.

PercyDan54 commented:
A suggestion: We might want to split the settings into two
Use seasonal background: Sometimes, always, never
Background Source: Beatmap, Skin

Commit:
Revert "Add BackgroundSource.Seasonal"

This reverts commit 2871001cc294da01f2db8e8c35d84a86ab1503ac.

Commit:
Revert "Set BackgroundSource.Seasonal as default setting"

This reverts commit cdb2d23578e6de7ca266a23a51b5fac2ed15b8f4.

Commit:
Revert "Load SeasonalBackgroundLoader asynchronously"

This reverts commit 81ebcd879668eb13cb28aa11bf4edfb8afb0fb99.

Commit:
Add separate Seasonal Backgrounds setting (Always, Sometimes, Never)

Commit:
Make SeasonalBackgroundLoader read from SessionStatics

Commit:
Load SeasonalBackgroundLoader asynchronously

Commit:
Show next Background on showSeasonalBackgrounds.ValueChanged

Commit:
Remove fallback texture parameter

 When there isn't a seasonal event, we don't want to fall back to the basic background here, but rather to the user selected background source.

Commit:
Load seasonal backgrounds according to setting

hbnrmx commented:
@bdach 
@PercyDan54 

Thanks for your feedback, and sorry I didn't get to this earlier.
In this iteration I
* split out Seasonal Backgrounds (Always/Sometimes/Never) from Background Source (Beatmap/Skin)
(I also feel this makes more sense in the case where no seasonal event is happening.)

* changed background fallback such that Beatmap/Skin source choice
 is respected when no Seasonal event is happening.

* made `SeasonalBackgroundLoader` read from `SessionStatics` 
(this should make sure that the API is queried only once for the index)

This still leaves:

* repeated texture fetching issue
* ugly `LoadComponentAsync` chain as described in the OP.

peppy commented:
"Sometimes" is based off whether we are currently "in season". This is provided by the api response [already](https://osu.ppy.sh/docs/index.html#apiv2seasonal-backgrounds) in the form of a timestamp (`ends_at`) so you should be able to consume this.

Commit:
Make "Sometimes" setting depend on season end date, rather than chance

peppy committed:
Merge branch 'master' into seasonal-backgrounds

peppy commented on osu.Game/Configuration/SessionStatics.cs:
```diff
@@ -12,12 +16,16 @@ protected override void InitialiseDefaults()
         {
             Set(Static.LoginOverlayDisplayed, false);
             Set(Static.MutedAudioNotificationShownOnce, false);
+            Set(Static.SeasonEndDate, DateTimeOffset.MinValue);
+            Set(Static.SeasonalBackgrounds, new List<APISeasonalBackground>());
```
Can probably just store the the whole response to one static?

peppy commented on osu.Game/Configuration/SeasonalBackgrounds.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace osu.Game.Configuration
+{
+    public enum SeasonalBackgrounds
```
`SeasonalBackgroundMode` will probably feel better

peppy commented on osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs:
```diff
@@ -102,6 +106,18 @@ private Background createBackground()
             else
                 newBackground = new Background(backgroundName);
 
+            switch (showSeasonalBackgrounds.Value)
+            {
+                case SeasonalBackgrounds.Sometimes:
+                    if (seasonalBackgroundLoader.IsInSeason())
+                        goto case SeasonalBackgrounds.Always;
+                    break;
+
+                case SeasonalBackgrounds.Always:
+                    newBackground = seasonalBackgroundLoader.LoadBackground() ?? newBackground;
+                    break;
+            }
```
this should probably be above everything else in this method. as it is, a different background is constructed above, then potentially thrown away due to the current sequence.

peppy commented on osu.Game/Screens/Backgrounds/BackgroundScreenDefault.cs:
```diff
@@ -102,6 +106,18 @@ private Background createBackground()
+            {
+                case SeasonalBackgrounds.Sometimes:
+                    if (seasonalBackgroundLoader.IsInSeason())
+                        goto case SeasonalBackgrounds.Always;
```
rather than using `goto`,  just write this as

`if (showSeasonalBackgrounds.Value == Always || (showSeasonalBackgrounds.Value == Sometimes))`

peppy commented on osu.Game/Graphics/Backgrounds/SeasonalBackgroundLoader.cs:
```diff
@@ -0,0 +1,76 @@
+            return new SeasonalBackground(url);
+        }
+
+        public bool IsInSeason() => DateTimeOffset.Now < endDate.Value;
```
```suggestion
        public bool IsInSeason => DateTimeOffset.Now < endDate.Value;
```

Commit:
Rename "SeasonalBackgrounds" to "SeasonalBackgroundMode"

Commit:
IsInSeason() -> IsInSeason

Commit:
Convert switch to if

Commit:
Save full api response in SessionStatics

Commit:
Move seasonalBackgroundMode check up and early return if available

Commit:
Fix nullref in case of successfull request but no backgrounds available

Commit:
Insert optional parentheses

bdach committed:
Fix nullref when querying IsInSeason before request completion

bdach committed:
Rename lookup & field

bdach committed:
Add xmldoc to SeasonalBackgroundMode

bdach committed:
Document nullability of seasonal backgrounds

bdach commented:
For the record I've fixed the test failures locally and am currently in the process of writing a test scene for this.

bdach committed:
Move config setting logic to background loader

bdach committed:
Add test coverage

bdach committed:
Allow to fetch if going online after launch

bdach committed:
Change background to seasonal right after login

bdach commented:
I've fixed the test failures, made a few stylistic changes, added test coverage, and also added support for delayed fetching of the backgrounds if API is not immediately available on launch, as the game would actually hard-crash if network was not available. The side effect of that change is that even though the API endpoint seems to be guest-accessible, the backgrounds won't be loaded until login, as we don't have an online-but-not-logged-in API state.

There's one transition at load that looks a little off that I want to look at but other than that I guess this is ok if we don't want to deal with caching the downloaded assets yet.

bdach committed:
Fix typo in comment

bdach committed:
Improve look & behaviour of background transitions

bdach commented:
It should be *sliiightly* better now. The fade in `SeasonalBackground.load` is a bit dodge but unsure if there's a better way to make sure it just doesn't come in without any transition at all and look absolutely janky.
